File: packages/sim/src/equipment.ts
import {WEAPON_PROFILES,weaponRangeText,type WeaponKind} from './weaponProfiles';
import {ammoUnit} from './flow';
import type { SimState } from './types';
import { nearDepot } from './flow';
import { take, drop, markShot } from './engineer';
import { heldItems } from './ledger';

import { isWeaponItem, type WeaponItem } from './weaponTypes';
export { isWeaponItem, type WeaponItem } from './weaponTypes';
export interface Weapon { kind: WeaponKind; loaded: number; cooldown: number; reload: number }
export interface Equipment {
  version: 1; next: number; weapons: Partial<Record<WeaponItem, Weapon>>;
  slots: [WeaponItem | null, WeaponItem | null]; active: 0 | 1;
  craft?: { seconds: number };
}
/** D-GP-START provisional first-cycle tuning; later paired-shot redesign stays GP-16. */
export const RIFLE = { steel: 10, copper: 4, seconds: 6, capacity: 10, reload: 1.5, rate: 2.5, range: WEAPON_PROFILES.rifle.effective, damage: 10 } as const;
export type EquipmentAction = { type: 'craftRifle' } | { type: 'cancelRifle' } | { type: 'reload' } | { type: 'swap' } | { type: 'equip'; item: WeaponItem; slot: 0 | 1 } | { type: 'unequip'; slot: 0 | 1 };
export function activeWeapon(st: SimState): Weapon | undefined {
  const q = st.engineer.equipment, id = q?.slots[q.active]; return id ? q!.weapons[id] : undefined;
}
function createWeapon(st: SimState, kind: Weapon['kind']): WeaponItem {
  const q = st.engineer.equipment!, id: WeaponItem = `${kind}:${q.next++}`;
  q.weapons[id] = { kind, loaded: 0, cooldown: 0, reload: 0 }; return id;
}
export function initEquipment(st: SimState, fresh = false): void {
  const e = st.engineer;
  if (!st.campaign?.progression?.gameplay || e.equipment) return;
  e.equipment = { version: 1, next: 1, weapons: {}, slots: [null, null], active: 0 };
  if (!fresh) {
    const id = createWeapon(st, e.barrels === 2 || st.campaign.progression.arsenal ? 'double' : 'rifle');
    e.equipment.slots[0] = id;
    const w = e.equipment.weapons[id]!;
    // Legacy fractional magazine remainder is loaded once; full magazines stay in the Backpack.
    const reserve = Math.round((e.inv.magazine ?? 0) * ammoUnit(st)), partial = reserve % 10;
    if (partial) { drop(e, 'magazine', partial / ammoUnit(st)); w.loaded = partial; }
    w.cooldown = Math.max(0, e.cooldown);
    st.flow!.stats.minedOf[id] = 1;
    st.campaign.progression.notice = 'Existing rifle capability preserved as equipped gear; ammunition conserved.';
  }
}
export function equipmentCheck(st: SimState, a: EquipmentAction): string {
  const e = st.engineer, q = e.equipment;
  if (!q) return 'Equipment belongs to the authored campaign';
  if (e.down >= 0 || e.truckSeat || st.campaign?.progression?.passenger) return 'Use equipment on foot';
  if (a.type === 'cancelRifle') return q.craft ? '' : 'No Rifle is being crafted';
  if (a.type === 'craftRifle') return !nearDepot(st) ? 'Walk to Home workshop' : q.craft ? 'Rifle crafting already in progress' :
    (e.inv.steel ?? 0) < RIFLE.steel || (e.inv.copper ?? 0) < RIFLE.copper ? 'Carry 10 Steel plates + 4 Copper' : '';
  if (a.type === 'equip' || a.type === 'unequip') {
    if (a.slot !== 0 && a.slot !== 1) return 'Choose equipment slot 1 or 2';
    if (a.type === 'equip' && (!isWeaponItem(a.item) || !q.weapons[a.item] || e.inv[a.item] !== 1)) return 'Select a carried weapon';
    const old = q.slots[a.slot];
    if (old) {
      const trial = { ...e, inv: { ...e.inv }, pack: e.pack?.map(s => s && { ...s }) };
      if (a.type === 'equip') drop(trial, a.item, 1);
      if (take(trial, old, 1) !== 1) return 'Make room in the Backpack for the equipped weapon';
    }
    return '';
  }
  if (a.type === 'swap') return q.slots[q.active === 0 ? 1 : 0] ? '' : 'Other equipment slot is empty';
  const w = activeWeapon(st);
  return !w ? 'Equip a Rifle' : w.reload > 0 ? 'Reloading' : w.loaded >= RIFLE.capacity ? 'Magazine is full' :
    Math.round((e.inv.magazine ?? 0) * ammoUnit(st)) <= 0 ? 'Carry bullets in your Backpack' : '';
}
export function equipmentCommand(st: SimState, a: EquipmentAction): { ok: boolean; reason: string } {
  const reason = equipmentCheck(st, a); if (reason) return { ok: false, reason };
  const e = st.engineer, q = e.equipment!;
  if (a.type === 'craftRifle') {
    for (const [item, n] of [['steel', RIFLE.steel], ['copper', RIFLE.copper]] as const) { drop(e, item, n); st.flow!.stats.consumed[item] += n; }
    q.craft = { seconds: 0 }; return { ok: true, reason: 'Crafting Rifle at Home · 6 seconds; output waits safely if Backpack is full' };
  }
  if (a.type === 'cancelRifle') { cancelRifleCraft(st); return { ok: true, reason: `Rifle crafting cancelled. ${RIFLE.steel} Steel plates + ${RIFLE.copper} Copper returned.` }; }
  if (a.type === 'reload') { activeWeapon(st)!.reload = RIFLE.reload; return { ok: true, reason: st.speed === 0 ? 'Reload queued · resume the game to load carried bullets' : 'Reloading · 1.5 seconds; uses carried bullets' }; }
  // Reload progress belongs to the weapon and pauses while it is stored.
  if (a.type === 'swap') q.active = q.active === 0 ? 1 : 0;
  else {
    const old = q.slots[a.slot];
    if (a.type === 'equip') drop(e, a.item, 1);
    if (old) take(e, old, 1);
    q.slots[a.slot] = a.type === 'equip' ? a.item : null;
    if (a.type === 'equip') q.active = a.slot;
  }
  e.cooldown = activeWeapon(st)?.cooldown ?? 0;
  return { ok: true, reason: 'Equipment updated; loaded ammunition and cooldown retained' };
}
/** GP-PLAYTEST-FIX 4 (2026-09-11): an unfinished Rifle returns its reserved plates to the Backpack (spilling to Home stock);
 *  the finished output is never refunded, so repeated cancellation cannot create resources. */
export function cancelRifleCraft(st: SimState): boolean {
  const e = st.engineer, q = e.equipment; if (!q?.craft) return false;
  for (const [item, n] of [['steel', RIFLE.steel], ['copper', RIFLE.copper]] as const) { const back = take(e, item, n); if (back < n) st.stock[item] += n - back; st.flow!.stats.consumed[item] -= n; }
  delete q.craft; return true;
}
export const rifleCrafting = (st: SimState): boolean => !!st.engineer.equipment?.craft;
export function tickEquipment(st: SimState, dt: number): void {
  const e = st.engineer, q = e.equipment; if (!q) return;
  if (q.craft && (e.down >= 0 || !nearDepot(st))) cancelRifleCraft(st);   // no remote or downed crafting: cancel and refund
  for (const w of Object.values(q.weapons)) if (w) w.cooldown = Math.max(0, w.cooldown - dt);
  if (q.craft && nearDepot(st) && e.down < 0) {
    q.craft.seconds = Math.min(RIFLE.seconds, q.craft.seconds + dt);
    if (q.craft.seconds >= RIFLE.seconds) {
      const id: WeaponItem = `rifle:${q.next}`;
      if (take(e, id, 1) === 1) { createWeapon(st, 'rifle'); st.flow!.stats.made[id] = 1; delete q.craft; }
    }
  }
  const w = activeWeapon(st);
  if (w && w.reload > 0 && e.down < 0 && !e.truckSeat && !st.campaign?.progression?.passenger) {
    w.reload = Math.max(0, w.reload - dt);
    if (w.reload === 0) { const n = Math.min(RIFLE.capacity - w.loaded, Math.round((e.inv.magazine ?? 0) * ammoUnit(st))); drop(e, 'magazine', n / ammoUnit(st)); w.loaded += n; }
  }
  e.cooldown = w?.cooldown ?? 0;
}
export function fireEquipment(st: SimState): boolean {
  const w = activeWeapon(st); if (!w || w.reload > 0 || w.cooldown > 1e-8 || w.loaded < 1) return false;
  w.loaded--; w.cooldown = 1 / WEAPON_PROFILES[w.kind].rate;
  st.engineer.fired++; if (st.engineer.firstShot < 0) st.engineer.firstShot = st.t;
  markShot(st); return true;
}
export function equipmentLabel(st: SimState): string {
  const q = st.engineer.equipment, w = activeWeapon(st);
  return !q ? 'Rifle' : !w ? Object.keys(q.weapons).length?'No weapon equipped · select your carried Rifle below, then Equip':'No weapon equipped · craft a Rifle at Home' :
    `${weaponRangeText(w.kind)} · ${w.loaded}/${RIFLE.capacity} loaded · ${Math.round((st.engineer.inv.magazine ?? 0) * ammoUnit(st))} reserve rounds${w.reload > 0 ? ` · Reload ${w.reload.toFixed(1)} s${st.speed === 0 ? ' · paused; resume to finish' : ''}` : ''}`;
}
export function equipmentProblem(st: SimState): string {
  const q = st.engineer.equipment; if (q === undefined) return '';
  if (!q || q.version !== 1 || !Number.isSafeInteger(q.next) || q.next < 1 || q.next > 999999999 ||
    !q.weapons || typeof q.weapons !== 'object' || Array.isArray(q.weapons) || !Array.isArray(q.slots) || q.slots.length !== 2 ||
    ![0, 1].includes(q.active) || q.slots.some(id => id !== null && (!isWeaponItem(id) || !q.weapons[id])) ||
    q.slots[0] !== null && q.slots[0] === q.slots[1] ||
    q.craft && (!Number.isFinite(q.craft.seconds) || q.craft.seconds < 0 || q.craft.seconds > RIFLE.seconds)) return 'Invalid equipment state';
  const total = heldItems(st).total;
  for (const [id, w] of Object.entries(q.weapons)) {
    if (!isWeaponItem(id) || Number(id.split(':')[1]) >= q.next || !w || w.kind !== id.split(':')[0] ||
      !Number.isSafeInteger(w.loaded) || w.loaded < 0 || w.loaded > RIFLE.capacity ||
      !Number.isFinite(w.cooldown) || w.cooldown < 0 || w.cooldown > 1 || !Number.isFinite(w.reload) || w.reload < 0 || w.reload > RIFLE.reload || total[id] !== 1) return 'Invalid weapon ownership or ammunition';
  }
  if (Object.entries(total).some(([id, n]) => isWeaponItem(id) && n > 0 && !q.weapons[id])) return 'Unknown weapon instance';
  return '';
}

File: Unity/Relight/Assets/Relight/Sim/Combat/Weapons/Equipment.cs
using System;
using System.Collections.Generic;

namespace Relight.Sim
{
    /// <summary>Craft a Rifle at the Home workshop (reference equipment.ts <c>{type:'craftRifle'}</c>).</summary>
    public sealed record CraftRifleCommand : Command;

    /// <summary>Abandon the Rifle craft; the reserved plates come back (reference <c>{type:'cancelRifle'}</c>).</summary>
    public sealed record CancelRifleCraftCommand : Command;

    /// <summary>Put a carried weapon in equipment slot 0 or 1 (reference <c>{type:'equip'}</c>).</summary>
    public sealed record EquipCommand(string Item, int Slot) : Command;

    /// <summary>Take the weapon out of slot 0 or 1 and back into the Backpack (reference <c>{type:'unequip'}</c>).</summary>
    public sealed record UnequipCommand(int Slot) : Command;

    /// <summary>Switch which equipment slot is in the hands (reference <c>{type:'swap'}</c>).</summary>
    public sealed record SwapWeaponCommand : Command;

    /// <summary>
    /// Put <paramref name="Key"/> on action-bar slot <paramref name="Slot"/> (0..9 = keys 1..0), or clear it with
    /// an empty key (UI_AND_ONBOARDING §5.5). A tool never sits on two slots: assigning one that is already
    /// elsewhere SWAPS the two slots rather than duplicating it.
    /// </summary>
    public sealed record AssignBarCommand(int Slot, string Key) : Command;

    /// <summary>A Rifle came off the Home workbench (presentation: the toast and the Backpack flash).</summary>
    public sealed record WeaponCraftedEvent(double T, string Item, string Kind) : SimEvent(T);

    /// <summary>Equipment changed: which slot, what is in it now ("" = empty) and which slot is in the hands.</summary>
    public sealed record EquipmentChangedEvent(double T, int Slot, string Item, int Active) : SimEvent(T);

    /// <summary>
    /// Owning, crafting, equipping and reloading weapons, ported from reference equipment.ts
    /// (<c>equipmentCheck</c> 42-64, <c>equipmentCommand</c> 65-85, <c>cancelRifleCraft</c> 88-92,
    /// <c>tickEquipment</c> 94-111, <c>fireEquipment</c> 112-117).
    ///
    /// Deliberate differences, each with its reason:
    /// * The reference gates every action on <c>st.campaign.progression.gameplay</c> and refuses with
    ///   "Equipment belongs to the authored campaign"; the port has no campaign progression yet (C-09) and the
    ///   equipment record always exists, so that gate is absent.
    /// * <c>truckSeat</c> / <c>passenger</c> are retired with the truck (CONTENT_CATALOGUE §17), so "on foot"
    ///   means "not down".
    /// * The reference's paused-game reload wording (<c>st.speed === 0</c>) is a session concern, not sim state.
    /// * A cancelled craft's plates that the Backpack cannot take yet become an explicit debt on
    ///   <see cref="WeaponState.CraftRefundSteel"/> rather than the retired Depot <c>st.stock</c>, exactly as
    ///   <see cref="HandCraft"/> does for a cancelled bullet batch — and <c>consumed</c> is un-counted only as the
    ///   debt is actually paid back, so the conservation ledger stays balanced at every tick without a new term.
    /// * Capacity, reload time and rate come from the weapon's own <see cref="WeaponDef"/> instead of the
    ///   reference's rifle-only <c>RIFLE</c> constants; for the rifle the numbers are identical (U-P-03).
    /// </summary>
    public static class WeaponRules
    {
        // ---- refusal texts, verbatim from reference equipment.ts:42-63 --------------------------------------
        public const string OnFootText = "Use equipment on foot";
        public const string NoCraftText = "No Rifle is being crafted";
        public const string WalkHomeText = "Walk to Home workshop";
        public const string BusyCraftText = "Rifle crafting already in progress";
        public const string SlotText = "Choose equipment slot 1 or 2";
        public const string CarriedText = "Select a carried weapon";
        public const string RoomText = "Make room in the Backpack for the equipped weapon";
        public const string OtherEmptyText = "Other equipment slot is empty";
        public const string NoWeaponText = "Equip a Rifle";
        public const string ReloadingText = "Reloading";
        public const string MagFullText = "Magazine is full";
        public const string NoBulletsText = "Carry bullets in your Backpack";
        public const string BarSlotText = "Choose an action bar slot";

        /// <summary>The rifle's recipe key in <see cref="GameData.Recipes"/> (catalogue row "rifle").</summary>
        public const string RifleRecipe = "rifle";

        /// <summary>
        /// The bullet row of <see cref="GameData.Ammunition"/> (reference flow.ts:156 <c>ammoUnit</c>: one item is
        /// <c>RoundsPerItem</c> rounds, always 1 per U-D-08). <see cref="WeaponDef"/> carries no ammunition key, so
        /// the port resolves it by the catalogue key rather than adding a field to a record another slot exports.
        /// </summary>
        public static AmmoDef Ammo(GameData d)
        {
            for (var i = 0; i < d.Ammunition.Count; i++)
                if (string.Equals(d.Ammunition[i].Key, "bullet", StringComparison.Ordinal)) return d.Ammunition[i];
            return d.Ammunition.Count > 0 ? d.Ammunition[0] : null;
        }

        /// <summary>Rounds the pockets can feed a magazine (reference <c>Math.round(inv.magazine * ammoUnit(st))</c>).</summary>
        public static double Reserve(GameData d, Engineer e)
        {
            var a = Ammo(d);
            return a == null ? 0 : Math.Round(e.Inv[a.Item] * a.RoundsPerItem);
        }

        private static bool OnFoot(SimState st) => !st.Engineer.IsDown;

        // ---- crafting --------------------------------------------------------------------------------------

        /// <summary>Reference <c>equipmentCheck</c> for <c>craftRifle</c> (47-48).</summary>
        public static string CheckCraft(SimContext ctx, SimState st)
        {
            if (!OnFoot(st)) return OnFootText;
            if (!HandCraft.NearDepot(ctx, st)) return WalkHomeText;
            if (st.Weapons.Crafting) return BusyCraftText;
            if (Home.RepairLocked(st)) return "Finish or cancel the repair first";
            var d = ctx.Data;
            if (!d.TryRecipe(RifleRecipe, out var r)) return "no such recipe";
            for (var i = 0; i < r.Inputs.Count; i++)
                if (st.Engineer.Inv[r.Inputs[i].Item] < r.Inputs[i].Count) return CarryText(d, r);
            return "";
        }

        /// <summary>Reference "Carry 10 Steel plates + 4 Copper", built from the recipe so tuning edits reach it.</summary>
        private static string CarryText(GameData d, Recipe r)
        {
            var s = "Carry";
            for (var i = 0; i < r.Inputs.Count; i++)
                s += (i == 0 ? " " : " + ") + r.Inputs[i].Count + " " + d.Item(r.Inputs[i].Item).DisplayName;
            return s;
        }

        /// <summary>Reference <c>equipmentCommand</c> 68-70: the plates are reserved up front and counted as consumed.</summary>
        public static (bool ok, string reason) Craft(SimContext ctx, SimState st)
        {
            var reason = CheckCraft(ctx, st);
            if (reason.Length > 0) return (false, reason);
            var d = ctx.Data;
            d.TryRecipe(RifleRecipe, out var r);
            for (var i = 0; i < r.Inputs.Count; i++)
            {
                Pockets.Drop(d, st.Engineer, ItemKey.Of(r.Inputs[i].Item), r.Inputs[i].Count);
                st.Stats.Consumed.Add(r.Inputs[i].Item, r.Inputs[i].Count);
            }
            st.Weapons.Crafting = true;
            st.Weapons.CraftSeconds = 0;
            return (true, $"Crafting {r.DisplayName} at the Home workshop · {PackLayout.Num(r.Seconds)} seconds. It keeps working while you are away; collect it from the workshop.");
        }

        /// <summary>
        /// Reference <c>cancelRifleCraft</c> (88-92). What the Backpack cannot take becomes a debt that
        /// <see cref="Drain"/> pays in as space appears; <c>consumed</c> is un-counted only for plates actually
        /// returned, so nothing is ever both un-counted and nowhere.
        /// </summary>
        public static bool CancelCraft(SimContext ctx, SimState st)
        {
            var q = st.Weapons;
            if (!q.Crafting) return false;
            var d = ctx.Data;
            d.TryRecipe(RifleRecipe, out var r);
            for (var i = 0; i < r.Inputs.Count; i++)
            {
                var owed = r.Inputs[i].Count;
                if (r.Inputs[i].Item == ItemId.Copper) q.CraftRefundCopper += owed;
                else q.CraftRefundSteel += owed;
            }
            q.Crafting = false;
            q.CraftSeconds = 0;
            Drain(d, st);
            return true;
        }

        /// <summary>Pays the cancelled craft's plates back into the pockets, as much as fits, and un-counts them.</summary>
        public static void Drain(GameData d, SimState st)
        {
            var q = st.Weapons;
            var e = st.Engineer;
            if (q.CraftRefundSteel > 0)
            {
                var back = Pockets.Take(d, e, ItemKey.Of(ItemId.Steel), q.CraftRefundSteel);
                q.CraftRefundSteel = Math.Max(0, q.CraftRefundSteel - back);
                st.Stats.Consumed.Add(ItemId.Steel, -back);
            }
            if (q.CraftRefundCopper > 0)
            {
                var back = Pockets.Take(d, e, ItemKey.Of(ItemId.Copper), q.CraftRefundCopper);
                q.CraftRefundCopper = Math.Max(0, q.CraftRefundCopper - back);
                st.Stats.Consumed.Add(ItemId.Copper, -back);
            }
        }

        /// <summary>Reference <c>createWeapon</c> (22-25): the next serial, the instance, and the id it is known by.</summary>
        public static string Create(SimState st, string kind)
        {
            var q = st.Weapons;
            var id = kind + ":" + q.Next.ToString(System.Globalization.CultureInfo.InvariantCulture);
            q.Next++;
            q.Owned.Add(new WeaponInstance { Id = id, Kind = kind, Loaded = 0, Cooldown = 0, Reload = 0 });
            return id;
        }

        // ---- equip / unequip / swap -------------------------------------------------------------------------

        /// <summary>Reference <c>equipmentCheck</c> 49-59 for <c>equip</c> / <c>unequip</c>.</summary>
        public static string CheckSlotChange(SimContext ctx, SimState st, string item, int slot, bool equip)
        {
            if (!OnFoot(st)) return OnFootText;
            if (slot != 0 && slot != 1) return SlotText;
            var d = ctx.Data;
            var e = st.Engineer;
            var q = st.Weapons;
            if (equip)
            {
                if (string.IsNullOrEmpty(item)) return CarriedText;
                var key = new ItemKey(item);
                if (!key.IsWeapon || q.Find(item) == null || e.Inv[key] != 1) return CarriedText;
            }
            var old = q.SlotAt(slot);
            if (!string.IsNullOrEmpty(old))
            {
                var trial = Pockets.Trial(e);
                if (equip) Pockets.Drop(d, trial, new ItemKey(item), 1);
                if (Pockets.Take(d, trial, new ItemKey(old), 1) != 1) return RoomText;
            }
            return "";
        }

        /// <summary>Reference <c>equipmentCommand</c> 76-84: the item leaves the pockets, the old one comes back.</summary>
        public static (bool ok, string reason) SlotChange(SimContext ctx, SimState st, string item, int slot, bool equip)
        {
            var reason = CheckSlotChange(ctx, st, item, slot, equip);
            if (reason.Length > 0) return (false, reason);
            var d = ctx.Data;
            var e = st.Engineer;
            var q = st.Weapons;
            var old = q.SlotAt(slot);
            if (equip) Pockets.Drop(d, e, new ItemKey(item), 1);
            if (!string.IsNullOrEmpty(old)) Pockets.Take(d, e, new ItemKey(old), 1);
            q.SetSlot(slot, equip ? item : null);
            if (equip) q.Active = slot;
            e.Cooldown = q.ActiveWeapon()?.Cooldown ?? 0;
            st.Events.Add(new EquipmentChangedEvent(st.T, slot, equip ? item : "", q.Active));
            return (true, "Equipment updated; loaded ammunition and cooldown retained");
        }

        /// <summary>Reference <c>equipmentCheck</c>/<c>equipmentCommand</c> for <c>swap</c> (60, 75).</summary>
        public static (bool ok, string reason) Swap(SimContext ctx, SimState st)
        {
            if (!OnFoot(st)) return (false, OnFootText);
            var q = st.Weapons;
            if (string.IsNullOrEmpty(q.SlotAt(q.Active == 0 ? 1 : 0))) return (false, OtherEmptyText);
            q.Active = q.Active == 0 ? 1 : 0;
            st.Engineer.Cooldown = q.ActiveWeapon()?.Cooldown ?? 0;
            st.Events.Add(new EquipmentChangedEvent(st.T, q.Active, q.SlotAt(q.Active) ?? "", q.Active));
            return (true, "Equipment updated; loaded ammunition and cooldown retained");
        }

        // ---- reload ------------------------------------------------------------------------------------------

        /// <summary>Reference <c>equipmentCheck</c> 61-63 (the fall-through case is <c>reload</c>).</summary>
        public static string CheckReload(SimContext ctx, SimState st)
        {
            if (!OnFoot(st)) return OnFootText;
            var w = st.Weapons.ActiveWeapon();
            if (w == null) return NoWeaponText;
            if (w.Reload > 0) return ReloadingText;
            var cap = Capacity(ctx.Data, w);
            if (w.Loaded >= cap) return MagFullText;
            if (Reserve(ctx.Data, st.Engineer) <= 0) return NoBulletsText;
            return "";
        }

        /// <summary>Reference <c>equipmentCommand</c> 73: the reload is started here and finishes in the tick.</summary>
        public static (bool ok, string reason) Reload(SimContext ctx, SimState st)
        {
            var reason = CheckReload(ctx, st);
            if (reason.Length > 0) return (false, reason);
            var w = st.Weapons.ActiveWeapon();
            var seconds = Seconds(ctx.Data, w);
            w.Reload = seconds;
            return (true, $"Reloading · {seconds.ToString("0.#", System.Globalization.CultureInfo.InvariantCulture)} seconds; uses carried bullets");
        }

        public static double Capacity(GameData d, WeaponInstance w) =>
            w != null && d.TryWeapon(w.Kind, out var p) ? p.Capacity : 0;

        public static double Seconds(GameData d, WeaponInstance w) =>
            w != null && d.TryWeapon(w.Kind, out var p) ? p.ReloadSeconds : 0;

        // ---- action bar --------------------------------------------------------------------------------------

        /// <summary>Action-bar slots, keys 1..9 then 0 (UI_AND_ONBOARDING §5.5).</summary>
        public const int BarSlots = 10;
        public static readonly string[] DefaultBar = { "generator", "excavator", "chest", "belt", "pole", "assembler", "turret", "lamp", "", "" };

        /// <summary>
        /// §5.5: one tool per slot and never the same tool twice. Assigning a key that already sits on another
        /// slot swaps the two slots; an empty key clears the slot.
        /// </summary>
        public static (bool ok, string reason) AssignBar(SimState st, int slot, string key)
        {
            if (slot < 0 || slot >= BarSlots) return (false, BarSlotText);
            var bar = st.Weapons.Bar;
            while (bar.Count < BarSlots) bar.Add("");
            key = key ?? "";
            var at = -1;
            for (var i = 0; i < BarSlots; i++) if (key.Length > 0 && string.Equals(bar[i], key, StringComparison.Ordinal)) { at = i; break; }
            if (at == slot) return (true, "");
            var displaced = bar[slot];
            bar[slot] = key;
            if (at >= 0) bar[at] = displaced;      // a swap, never a duplicate
            return (true, "");
        }

        /// <summary>A loaded arrangement with a duplicate collapses to the first occurrence (§5.5).</summary>
        public static void NormaliseBar(SimState st)
        {
            var bar = st.Weapons.Bar;
            while (bar.Count < BarSlots) bar.Add("");
            if (bar.Count > BarSlots) bar.RemoveRange(BarSlots, bar.Count - BarSlots);
            for (var i = 0; i < bar.Count; i++)
            {
                if (bar[i] == null) { bar[i] = ""; continue; }
                if (bar[i].Length == 0) continue;
                for (var j = 0; j < i; j++) if (string.Equals(bar[j], bar[i], StringComparison.Ordinal)) { bar[i] = ""; break; }
            }
        }
    }

    public sealed class EquipmentHandler : ICommandHandler
    {
        public bool TryApply(SimContext ctx, SimState st, Command c, out CommandResult result)
        {
            switch (c)
            {
                case CraftRifleCommand _:
                {
                    var (ok, reason) = WeaponRules.Craft(ctx, st);
                    result = ok ? CommandResult.Ok(reason) : CommandResult.Refuse(reason);
                    return true;
                }
                case CancelRifleCraftCommand _:
                {
                    if (!st.Weapons.Crafting) { result = CommandResult.Refuse(WeaponRules.NoCraftText); return true; }
                    ctx.Data.TryRecipe(WeaponRules.RifleRecipe, out var r);
                    WeaponRules.CancelCraft(ctx, st);
                    var back = "";
                    for (var i = 0; i < r.Inputs.Count; i++)
                        back += (i == 0 ? " " : " + ") + r.Inputs[i].Count + " " + ctx.Data.Item(r.Inputs[i].Item).DisplayName;
                    result = CommandResult.Ok($"{r.DisplayName} crafting cancelled.{back} returned.");
                    return true;
                }
                case EquipCommand e:
                {
                    var (ok, reason) = WeaponRules.SlotChange(ctx, st, e.Item, e.Slot, true);
                    result = ok ? CommandResult.Ok(reason) : CommandResult.Refuse(reason);
                    return true;
                }
                case UnequipCommand u:
                {
                    var (ok, reason) = WeaponRules.SlotChange(ctx, st, null, u.Slot, false);
                    result = ok ? CommandResult.Ok(reason) : CommandResult.Refuse(reason);
                    return true;
                }
                case SwapWeaponCommand _:
                {
                    var (ok, reason) = WeaponRules.Swap(ctx, st);
                    result = ok ? CommandResult.Ok(reason) : CommandResult.Refuse(reason);
                    return true;
                }
                case AssignBarCommand a:
                {
                    var (ok, reason) = WeaponRules.AssignBar(st, a.Slot, a.Key);
                    result = ok ? CommandResult.Ok() : CommandResult.Refuse(reason);
                    return true;
                }
                default:
                    result = default;
                    return false;
            }
        }
    }
}
